proparse/src/main/java/org/prorefactor/proparse/antlr4/Lexer.java

```diff
@@ -0,0 +1,1117 @@
+/*******************************************************************************
+ * Copyright (c) 2003-2015 John Green
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    John Green - initial API and implementation and/or initial documentation
+ *******************************************************************************/ 
+package org.prorefactor.proparse.antlr4;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+
+import org.prorefactor.macrolevel.MacroDef;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Lexer  {
+  private static final Logger LOGGER = LoggerFactory.getLogger(Lexer.class);
+
+  private static final int EOF_CHAR = -1;
+
+  /** Lowercase value of current character */
+  private int currChar;
+
+  /** Current character, before being lowercased */
+  private int currInt;
+  private int currFile, currLine, currCol;
+  private int prevFile, prevLine, prevCol;
+  
+  private int currStringType;
+  private StringBuilder currText = new StringBuilder();
+
+  private ProgressLexer prepro;
+
+  private boolean gettingAmpIfDefArg = false;
+  private boolean preserve = false;
+  private int preserveFile;
+  private int preserveLine;
+  private int preserveCol;
+  private int preserveSource;
+  private int preserveChar;
+
+  private int textStartFile;
+  private int textStartLine;
+  private int textStartCol;
+  private int textStartSource;
+
+  private Set<Integer> comments = new HashSet<>();
+  private Set<Integer> loc = new HashSet<>();
+
+  Lexer(ProgressLexer prepro) throws IOException {
+    this.prepro = prepro;
+    getChar(); // We always assume "currChar" is available.
+  }
+
+
+  //////////////// Lexical productions listed first, support functions follow.
+  public ProToken nextToken() {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "nextToken" is 59 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)